### PR TITLE
feat(components): KFP - Run component

### PR DESCRIPTION
**Pull request by Ark-kun**
This can be useful for dynamically running components or pipelines.

### CODE REVIEW

File: components/kfp/Run_component/component.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+) -> NamedTuple('Outputs', [
+    ('run_id', str),
+    ('run_object', 'JsonObject'),  # kfp.ApiRunDetails
+]):
+    import json
+    import os
+    import kfp
+    from kfp_server_api import ApiClient
+    print('Loading component...')
+    op = kfp.components.load_component_from_url(component_url)
```
Comment (Bobgy) [inline components/kfp/Run_component/component.py:18]: This looks awesome! I'd like to use this component for https://github.com/kubeflow/pipelines/pull/5337.

However, how does this support running a pipeline? graph component?

After thinking about my use-case, I think a few options would enable my scenario:
1. I want to load a component/pipeline from local path too (passed from input artifacts).
2. I want to get run status also as an explicit output parameter.
3. the script can get very complex with many options, consider always adding unit tests next to a component?

What do you think? Note, I might just go ahead and implement a fork of this component to satisfy immediate needs quickly, but I think it'll be worth it standardizing what we do for our own tests as a reusable component that everyone can use to set up their pipeline CI/CD.

File: components/kfp/Run_component/component.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        run_id,
+        json.dumps(run_dict, indent=4),
+    )
+
+
+if __name__ == '__main__':
+    from kfp.components import create_component_from_func
+    run_component_or_pipeline_op = create_component_from_func(
+        run_component_or_pipeline,
+        base_image='python:3.9',
```
Comment (Bobgy) [inline components/kfp/Run_component/component.py:41]: shall we use a smaller base_image like `alpine:latest`? alpine also comes with python

File: components/kfp/Run_component/component.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    if not endpoint:
+        endpoint = os.environ['ML_PIPELINE_SERVICE_HOST'] + ':' + os.environ['ML_PIPELINE_SERVICE_PORT']
+    create_run_result = kfp.Client(host=endpoint).create_run_from_pipeline_func(op, arguments=arguments)
+    run_id = str(create_run_result.run_id)
+    print('Submitted run: ' + run_id)
+    print('Waiting for the run to finish...')
+    run_object = create_run_result.wait_for_run_completion(wait_timeout_seconds)
+    print('Run has finished.')
+    # sanitize_for_serialization uses correct field names and properly converts datetime values
+    run_dict = ApiClient().sanitize_for_serialization(run_object)
```
Comment (Bobgy) [inline components/kfp/Run_component/component.py:30]: what does this do?

File: components/kfp/Run_component/component.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    print('Loading component...')
+    op = kfp.components.load_component_from_url(component_url)
+    print('Loading component done.')
+    print('Submitting run...')
+    if not endpoint:
+        endpoint = os.environ['ML_PIPELINE_SERVICE_HOST'] + ':' + os.environ['ML_PIPELINE_SERVICE_PORT']
+    create_run_result = kfp.Client(host=endpoint).create_run_from_pipeline_func(op, arguments=arguments)
+    run_id = str(create_run_result.run_id)
+    print('Submitted run: ' + run_id)
+    print('Waiting for the run to finish...')
```
Comment (Bobgy) [inline components/kfp/Run_component/component.py:26]: I think it'd be very useful emitting another output, which is UI url of the pipeline run details page.

File: components/kfp/Run_component/component.py (hunk 1 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:18]: Thank you =)

>However, how does this support running a pipeline? graph component?

Graph component is an option.
But I also think we should add more `kfp` components alongside. For example, a component to submit a pipeline from python code. `Run_pipeline`.

>I want to load a component/pipeline from local path too (passed from input artifacts).

Good idea. I should probably split this component into `from_file` and `from_url`.

>I want to get run status also as an explicit output parameter.

Good idea. I'll add that.

>the script can get very complex with many options, consider always adding unit tests next to a component?

What would be an action item here? I usually add sample pipelines for every component or a group of components.

File: components/kfp/Run_component/component.py (hunk 4 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:26]: How would you build that URL?

File: components/kfp/Run_component/component.py (hunk 3 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:30]: This is basically `.to_dict()` that works properly (`camelCase`, not `snake_case`). I thought the issue was fixed in the Swagger/OpenAPI generator, but maybe has it resurfaced.

File: components/kfp/Run_component/component.py (hunk 2 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:41]: Are you sure? `alpine` does not seem to have Python. It does not even have `bash`... That's why it's small.

File: components/kfp/Run_component/component.py (hunk 4 above)
Comment (Bobgy) [reply components/kfp/Run_component/component.py:26]: https://github.com/kubeflow/pipelines/blob/fc769b82a752b5680a524bd4e488e634ef04aa23/v2/test/samples/fail.py#L48

I pass the real hostname into the component

File: components/kfp/Run_component/component.py (hunk 3 above)
Comment (Bobgy) [reply components/kfp/Run_component/component.py:30]: I'm a little confused, is `camelCase` the expected output? why?

File: components/kfp/Run_component/component.py (hunk 1 above)
Comment (Bobgy) [reply components/kfp/Run_component/component.py:18]: > For example, a component to submit a pipeline from python code. Run_pipeline.

I think there are a lot of user interest for this too.

> Good idea. I should probably split this component into from_file and from_url.

I'd prefer sharing most of the code, but allow two optional inputs, one for from file and one for from url. One single component that works for many scenarios sound easier to understand -- think about unix philosophy of building tools that are small, focused and does a lot of things flexibly. It's harder to manage a larger set of tools.

Comment (Bobgy) [reply components/kfp/Run_component/component.py:18]: my example: https://github.com/kubeflow/pipelines/pull/5337/files#diff-9cf6e1982d0f982025dd3d53d87a82aa46b313b78511769cf66629aa22e36fbdR4-R15

do you think that's easier to understand than two separate components?

Comment (Bobgy) [reply components/kfp/Run_component/component.py:18]: > What would be an action item here? I usually add sample pipelines for every component or a group of components.

AI: add presubmit test that runs unit test for all component unit tests. I think it'll be great if we build components following all the engineering best practices -- including CI.

Comment (Bobgy) [reply components/kfp/Run_component/component.py:18]: One thing I don't have a good idea right now is that, I need to set component image dynamically and use a pipeline from a local path: https://github.com/kubeflow/pipelines/blob/fc769b82a752b5680a524bd4e488e634ef04aa23/v2/test/sample_test.py#L50

How do I do that with this component? or what pattern should I switch to?

File: components/kfp/Run_component/component.py (hunk 3 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:30]: That's what Kubernetes REST API expects. AFAIK, the server will reject camel_case attributes.
But some Swagger generated clients have a bug and uses python names in `.to_dict()`.

File: components/kfp/Run_component/component.py (hunk 1 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py:18]: >AI: add presubmit test that runs unit test for all component unit tests. I think it'll be great if we build components following all the engineering best practices -- including CI.

Sure. Let's start doing that once your new sample test infra gets in.

Comment (Ark-kun) [reply components/kfp/Run_component/component.py:18]: >I'd prefer sharing most of the code, but allow two optional inputs, one for from file and one for from url.

This does not always work (e.g. for TFX components), but I guess for this component it's OK.

P.S. It's a bit confusing when optional inputs are not really optional (one must be provided).

**Comment by Ark-kun** [comment]
Let's get this in and do incremental changes as needed.

File: components/kfp/Run_component/_samples/sample_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from kfp import components
+
+kfp_endpoint = None
+
+
+run_component_or_pipeline_op = components.load_component_from_url('https://raw.githubusercontent.com/kubeflow/pipelines/2d13b358690c761f64e5b59b70624de8f1f52a29/components/kfp/Run_component/component.yaml')
+
+
+def my_pipeline():
```
Comment (Bobgy) [inline components/kfp/Run_component/_samples/sample_pipeline.py:9]: Can we move the sample to samples top folder and add a test in the new sample test infra?

You can add a readme with link to the sample.

**Review by Bobgy — COMMENTED** [review]
Besides that, can you remind me why it's named Run_component?

With the two issues solved, I think it'll be ready to merge

**Comment by Ark-kun** [comment]
>Besides that, can you remind me why it's named Run_component?

What name would you propose?
This component launches another component (container or graph). It launches a component, not a python pipeline or compiled pipeline.

File: components/kfp/Run_component/component.py (hunk 4 above)
Comment (Ark-kun) [reply components/kfp/Run_component/component.py]: Done

File: components/kfp/Run_component/_samples/sample_pipeline.py (hunk 5 above)
Comment (Ark-kun) [reply components/kfp/Run_component/_samples/sample_pipeline.py:9]: I've created a feature request to add the sample to the tests: https://github.com/kubeflow/pipelines/issues/5755

**Comment by Bobgy** [comment]
@Ark-kun I was talking about the start with uppercase, but use underscore to connect words convention. It's a very uncommon naming pattern as far as I tell. Shall we consider CamelCase or snake_case or resource-name like styles instead?
anyway that's not a blocker

**Pull request merged**